File: ui/android/java/src/org/chromium/ui/gl/SurfaceTextureListener.java
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.ui.gl;

import android.graphics.SurfaceTexture;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.MainDex;
import org.chromium.base.annotations.NativeMethods;

/**
 * Listener to an android SurfaceTexture object for frame availability.
 */
@JNINamespace("gl")
@MainDex
class SurfaceTextureListener implements SurfaceTexture.OnFrameAvailableListener {
    // Used to determine the class instance to dispatch the native call to.
    private final long mNativeSurfaceTextureListener;

    SurfaceTextureListener(long nativeSurfaceTextureListener) {
        assert nativeSurfaceTextureListener != 0;
        mNativeSurfaceTextureListener = nativeSurfaceTextureListener;
    }

    @Override
    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
        SurfaceTextureListenerJni.get().frameAvailable(
                mNativeSurfaceTextureListener, SurfaceTextureListener.this);
    }

    @Override
    protected void finalize() throws Throwable {
        try {
            SurfaceTextureListenerJni.get().destroy(
                    mNativeSurfaceTextureListener, SurfaceTextureListener.this);
        } finally {
            super.finalize();
        }
    }

    @NativeMethods
    interface Natives {
        void frameAvailable(long nativeSurfaceTextureListener, SurfaceTextureListener caller);
        void destroy(long nativeSurfaceTextureListener, SurfaceTextureListener caller);
    }
}

File: ui/gl/android/surface_texture_listener.cc
// Copyright 2013 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "ui/gl/android/surface_texture_listener.h"

#include <utility>

#include "base/location.h"
#include "base/task/single_thread_task_runner.h"

namespace gl {

SurfaceTextureListener::SurfaceTextureListener(base::RepeatingClosure callback,
                                               bool use_any_thread)
    : callback_(std::move(callback)),
      browser_loop_(base::SingleThreadTaskRunner::GetCurrentDefault()),
      use_any_thread_(use_any_thread) {}

SurfaceTextureListener::~SurfaceTextureListener() {
}

void SurfaceTextureListener::Destroy(JNIEnv* env) {
  if (!browser_loop_->DeleteSoon(FROM_HERE, this)) {
    delete this;
  }
}

void SurfaceTextureListener::FrameAvailable(JNIEnv* env) {
  if (!use_any_thread_ && !browser_loop_->BelongsToCurrentThread()) {
    browser_loop_->PostTask(FROM_HERE, callback_);
  } else {
    callback_.Run();
  }
}

}  // namespace gl
